<center>
<h4>Diplomatura en CDAAyA 2018 - FaMAF - UNC</h4>
<h2>Laboratorio de Exploración y Curación de Datos</h2>
<h3>Limpieza de Base de Permisos de Cosntrucción en San Francisco</h3>
</center>
</left>
<h4>Lucas Prone y Julieta Bergamasco</h4>
</left>

## Introducción

En la siguiente notebook, implementaremos los pasos necesarios para la limpieza de una base de datos de permisos de construcción en San Francisco, EEUU. Para ello, seguiremos el *check list* propuesto en clase.

## 1. Importación de Datos

### 1.1. Verificación de Inexistencia de Problemas en la Importación

Para comenzar, importamos los datos que vamos a procesar:

In [1]:
import pandas as pd

In [2]:
#data = pd.read_csv('Building_Permits.csv',low_memory=False)
data = pd.read_csv('https://github.com/sac2087/famaf/blob/master/Analisis%20exploratorio%20y%20curacion%20de%20datos/Ejercicios/Building_Permits.csv?raw=true',low_memory=False)

In [3]:
pd.set_option('display.max_columns', None)
data.head(5)

Permit Number  Permit Type            Permit Type Definition  \
0  201505065519            4                      sign - erect   
1  201604195146            4                      sign - erect   
2  201605278609            3  additions alterations or repairs   
3  201611072166            8            otc alterations permit   
4  201611283529            6                       demolitions   

  Permit Creation Date Block  Lot  Street Number Street Number Suffix  \
0           05/06/2015  0326  023            140                  NaN   
1           04/19/2016  0306  007            440                  NaN   
2           05/27/2016  0595  203           1647                  NaN   
3           11/07/2016  0156  011           1230                  NaN   
4           11/28/2016  0342  001            950                  NaN   

  Street Name Street Suffix  Unit Unit Suffix  \
0       Ellis            St   NaN         NaN   
1       Geary            St   0.0         NaN   
2     Pacific            Av   NaN         NaN   
3     Pacific            Av   0.0         NaN   
4      Market            St   NaN         NaN   

                                         Description Current Status  \
0  ground fl facade: to erect illuminated, electr...        expired   
1            remove (e) awning and associated signs.         issued   
2                    installation of separating wall      withdrawn   
3           repair dryrot & stucco at front of bldg.       complete   
4  demolish retail/office/commercial 3-story buil...         issued   

  Current Status Date  Filed Date Issued Date Completed Date  \
0          12/21/2017  05/06/2015  11/09/2015            NaN   
1          08/03/2017  04/19/2016  08/03/2017            NaN   
2          09/26/2017  05/27/2016         NaN            NaN   
3          07/24/2017  11/07/2016  07/18/2017     07/24/2017   
4          12/01/2017  11/28/2016  12/01/2017            NaN   

  First Construction Document Date Structural Notification  \
0                       11/09/2015                     NaN   
1                       08/03/2017                     NaN   
2                              NaN                     NaN   
3                       07/18/2017                     NaN   
4                       11/20/2017                     NaN   

   Number of Existing Stories  Number of Proposed Stories  \
0                         6.0                         NaN   
1                         7.0                         NaN   
2                         6.0                         6.0   
3                         2.0                         2.0   
4                         3.0                         NaN   

  Voluntary Soft-Story Retrofit Fire Only Permit Permit Expiration Date  \
0                           NaN              NaN             11/03/2016   
1                           NaN              NaN             12/03/2017   
2                           NaN              NaN                    NaN   
3                           NaN              NaN             07/13/2018   
4                           NaN              NaN             12/01/2018   

   Estimated Cost  Revised Cost         Existing Use  Existing Units  \
0          4000.0        4000.0  tourist hotel/motel           143.0   
1             1.0         500.0  tourist hotel/motel             NaN   
2         20000.0           NaN         retail sales            39.0   
3          2000.0        2000.0    1 family dwelling             1.0   
4        100000.0      100000.0         retail sales             NaN   

        Proposed Use  Proposed Units  Plansets TIDF Compliance  \
0                NaN             NaN       2.0             NaN   
1                NaN             NaN       2.0             NaN   
2       retail sales            39.0       2.0             NaN   
3  1 family dwelling             1.0       2.0             NaN   
4                NaN             NaN       2.0             NaN   

   Existing Construction Type Existing

Tomamos una muestra aleatoria para ver valores más dispersos.

In [4]:
# set seed for reproducibility
import numpy as np
np.random.seed(0)
data.sample(20)

Permit Number  Permit Type            Permit Type Definition  \
40553   201403039652            8            otc alterations permit   
169731  201510159735            3  additions alterations or repairs   
19180        M409787            8            otc alterations permit   
68047   201411191888            8            otc alterations permit   
64238        M527228            8            otc alterations permit   
27703        M434527            8            otc alterations permit   
109281  201512175360            8            otc alterations permit   
15075   201306109020            8            otc alterations permit   
59124   201408264765            8            otc alterations permit   
175167  201702240086            8            otc alterations permit   
157845  201705247586            8            otc alterations permit   
150068       M762087            8            otc alterations permit   
169291  201708164983            8            otc alterations permit   
61608   201409196774            3  additions alterations or repairs   
34145   201312174422            8            otc alterations permit   
181163  201709067013            8            otc alterations permit   
593          M364468            8            otc alterations permit   
198604  201802221963            8            otc alterations permit   
117133       M671487            8            otc alterations permit   
166700       M815987            8            otc alterations permit   

       Permit Creation Date Block   Lot  Street Number Street Number Suffix  \
40553            03/03/2014  3732   008            400                  NaN   
169731           10/15/2015  2609   028             79                  NaN   
19180            07/22/2013  4624   031            178                  NaN   
68047            11/19/2014  0039   109            294                  NaN   
64238            10/14/2014  1251   002            707                  NaN   
27703            10/11/2013  0348   015            277                  NaN   
109281           12/17/2015  0061   013             40                  NaN   
15075            06/10/2013  0262   020            100                  NaN   
59124            08/26/2014  0321   016            650                  NaN   
175167           02/24/2017  6533   028            451                  NaN   
157845           05/24/2017  2630   018           1171                  NaN   
150068           02/17/2017  1776  009D           1355                  NaN   
169291           08/16/2017  2347  004G           2350                  NaN   
61608            09/19/2014  2705   029             98                  NaN   
34145            12/17/2013  3514   030           1663                  NaN   
181163           09/06/2017  0289   005            100                  NaN   
593              01/03/2013  0604   023           2330                  NaN   
198604           02/22/2018  0931   013           2544                  NaN   
117133           03/10/2016  5651   027            113                  NaN   
166700           08/03/2017  0868   066            108                    B   

        Street Name Street Suffix  Unit Unit Suffix  \
40553    Clementina            St   NaN         NaN   
169731  Buena Vista            Tr   NaN         NaN   
19180    West Point            Rd   NaN         NaN   
68047     Francisco            St   0.0         NaN   
64238          Cole            St   NaN         NaN   
27703   Golden Gate            Av   0.0         NaN   
109281      Whiting            St   NaN         NaN   
15075          Pine            St   0.0         NaN   
59124         Ellis            St   NaN         NaN   
175167    Fair Oaks            St   NaN         NaN   
157845      Stanyan            St   NaN         NaN   
150068         21st            Av   NaN         NaN   
169291         19th            Av   NaN         NaN   
61608         Crown            Tr   NaN         NaN   
34145       Mission           

Veamos los tipos de datos que tenemos

In [9]:
data_types = data.dtypes
data_types = pd.DataFrame(data_types)
data_types = data_types.rename(columns = {0:'ReadingDataType'})
data_types.index.names = ['FieldName']
data_types

ReadingDataType
FieldName                                             
Permit Number                                   object
Permit Type                                      int64
Permit Type Definition                          object
Permit Creation Date                            object
Block                                           object
Lot                                             object
Street Number                                    int64
Street Number Suffix                            object
Street Name                                     object
Street Suffix                                   object
Unit                                           float64
Unit Suffix                                     object
Description                                     object
Current Status                                  object
Current Status Date                             object
Filed Date                                      object
Issued Date                                     object
Completed Date                                  object
First Construction Document Date                object
Structural Notification                         object
Number of Existing Stories                     float64
Number of Proposed Stories                     float64
Voluntary Soft-Story Retrofit                   object
Fire Only Permit                                object
Permit Expiration Date                          object
Estimated Cost                                 float64
Revised Cost                                   float64
Existing Use                                    object
Existing Units                                 float64
Proposed Use                                    object
Proposed Units                                 float64
Plansets                                       float64
TIDF Compliance                                 object
Existing Construction Type                     float64
Existing Construction Type Description          object
Proposed Construction Type                     float64
Proposed Construction Type Description          object
Site Permit                                     object
Supervisor District                            float64
Neighborhoods - Analysis Boundaries             object
Zipcode                                        float64
Location                                        object
Record ID                                        int64

Según la metadata que acompaña el dataset, tenemos lo siguiente:

In [10]:
data_def_types = pd.DataFrame([["Permit Number","String"],
["Permit Type","Numeric"],
["Permit Type Definition","String"],
["Permit Creation Date","DateTime"],
["Block","String"],
["Lot","String"],
["Street Number","Numeric"],
["Street Number Suffix","String"],
["Street Name","String"],
["Street Suffix","String"],
["Unit","String"],
["Unit Suffix","String"],
["Description","String"],
["Current Status","String"],
["Current Status Date","DateTime"],
["Filed Date","DateTime"],
["Issued Date","String"],
["Completed Date","String"],
["First Construction Document Date","String"],
["Structural Notification","String"],
["Number of Existing Stories","String"],
["Number of Proposed Stories","String"],
["Voluntary Soft-Story Retrofit","String"],
["Fire Only Permit","String"],
["Permit Expiration Date","String"],
["Estimated Cost","String"],
["Revised Cost","String"],
["Existing Use","String"],
["Existing Units","String"],
["Proposed Use","String"],
["Proposed Units","String"],
["Plansets","String"],
["TIDF Compliance","String"],
["Existing Construction Type","String"],
["Existing Construction Type Description","String"],
["Proposed Construction Type","String"],
["Proposed Construction Type Description","String"],
["Site Permit","String"],
["Supervisor District","Numeric"],
["Neighborhoods - Analysis Boundaries","String"],
["Zipcode","Numeric"],
["Location","String"],
["Record ID","Numeric"]],
columns=["FieldName","ProvidedType"])

data_def_types = data_def_types.set_index("FieldName")
data_def_types

ProvidedType
FieldName                                          
Permit Number                                String
Permit Type                                 Numeric
Permit Type Definition                       String
Permit Creation Date                       DateTime
Block                                        String
Lot                                          String
Street Number                               Numeric
Street Number Suffix                         String
Street Name                                  String
Street Suffix                                String
Unit                                         String
Unit Suffix                                  String
Description                                  String
Current Status                               String
Current Status Date                        DateTime
Filed Date                                 DateTime
Issued Date                                  String
Completed Date                               String
First Construction Document Date             String
Structural Notification                      String
Number of Existing Stories                   String
Number of Proposed Stories                   String
Voluntary Soft-Story Retrofit                String
Fire Only Permit                             String
Permit Expiration Date                       String
Estimated Cost                               String
Revised Cost                                 String
Existing Use                                 String
Existing Units                               String
Proposed Use                                 String
Proposed Units                               String
Plansets                                     String
TIDF Compliance                              String
Existing Construction Type                   String
Existing Construction Type Description       String
Proposed Construction Type                   String
Proposed Construction Type Description       String
Site Permit                                  String
Supervisor District                         Numeric
Neighborhoods - Analysis Boundaries          String
Zipcode                                     Numeric
Location                                     String
Record ID                                   Numeric

Comparando, tenemos lo sigiente:

In [11]:
pd.merge(data_types, data_def_types, left_index=True, right_index=True)

ReadingDataType ProvidedType
FieldName                                                          
Permit Number                                   object       String
Permit Type                                      int64      Numeric
Permit Type Definition                          object       String
Permit Creation Date                            object     DateTime
Block                                           object       String
Lot                                             object       String
Street Number                                    int64      Numeric
Street Number Suffix                            object       String
Street Name                                     object       String
Street Suffix                                   object       String
Unit                                           float64       String
Unit Suffix                                     object       String
Description                                     object       String
Current Status                                  object       String
Current Status Date                             object     DateTime
Filed Date                                      object     DateTime
Issued Date                                     object       String
Completed Date                                  object       String
First Construction Document Date                object       String
Structural Notification                         object       String
Number of Existing Stories                     float64       String
Number of Proposed Stories                     float64       String
Voluntary Soft-Story Retrofit                   object       String
Fire Only Permit                                object       String
Permit Expiration Date                          object       String
Estimated Cost                                 float64       String
Revised Cost                                   float64       String
Existing Use                                    object       String
Existing Units                                 float64       String
Proposed Use                                    object       String
Proposed Units                                 float64       String
Plansets                                       float64       String
TIDF Compliance                                 object       String
Existing Construction Type                     float64       String
Existing Construction Type Description          object       String
Proposed Construction Type                     float64       String
Proposed Construction Type Description          object       String
Site Permit                                     object       String
Supervisor District                            float64      Numeric
Neighborhoods - Analysis Boundaries             object       String
Zipcode                                        float64      Numeric
Location                                        object       String
Record ID                                        int64      Numeric

Como puede verse en la comparación, algunos campos originalmente tipificados como `String` han sido identificados como numéricos, pero esto no generará inconvenientes. Además todos los campos numéricos han sido catalogados como `int` o `float` correctamente. Sin embargo, las columnas **Permit Creation Date**, **Current Status Date** y **Filed Date** han sido identificadas como `object`, pero contienen datos del tipo `DateTime`.

Para salvar este error de tipificación, realizamos nuevamente la lectura, especificando el tipo de datos para estas columnas:

In [12]:
data = pd.read_csv('Building_Permits.csv',low_memory=False,
                   parse_dates=["Permit Creation Date","Current Status Date","Filed Date"])

In [13]:
data.head(10)

Permit Number  Permit Type            Permit Type Definition  \
0  201505065519            4                      sign - erect   
1  201604195146            4                      sign - erect   
2  201605278609            3  additions alterations or repairs   
3  201611072166            8            otc alterations permit   
4  201611283529            6                       demolitions   
5  201706149344            8            otc alterations permit   
6  201706300814            8            otc alterations permit   
7       M803667            8            otc alterations permit   
8       M804227            8            otc alterations permit   
9       M804767            8            otc alterations permit   

  Permit Creation Date Block  Lot  Street Number Street Number Suffix  \
0           2015-05-06  0326  023            140                  NaN   
1           2016-04-19  0306  007            440                  NaN   
2           2016-05-27  0595  203           1647                  NaN   
3           2016-11-07  0156  011           1230                  NaN   
4           2016-11-28  0342  001            950                  NaN   
5           2017-06-14  4105  009            800                  NaN   
6           2017-06-30  1739  020           1291                  NaN   
7           2017-06-30  4789  014           1465                  NaN   
8           2017-07-05  1212  054           2094                  NaN   
9           2017-07-06  1259  016             89                  NaN   

  Street Name Street Suffix  Unit Unit Suffix  \
0       Ellis            St   NaN         NaN   
1       Geary            St   0.0         NaN   
2     Pacific            Av   NaN         NaN   
3     Pacific            Av   0.0         NaN   
4      Market            St   NaN         NaN   
5     Indiana            St   NaN         NaN   
6        11th            Av   0.0         NaN   
7      Revere            Av   NaN         NaN   
8        Fell            St   NaN         NaN   
9      Alpine            Tr   NaN         NaN   

                                         Description Current Status  \
0  ground fl facade: to erect illuminated, electr...        expired   
1            remove (e) awning and associated signs.         issued   
2                    installation of separating wall      withdrawn   
3           repair dryrot & stucco at front of bldg.       complete   
4  demolish retail/office/commercial 3-story buil...         issued   
5                                          evac maps         issued   
6                                         re-roofing       complete   
7                                        streetspace         issued   
8                                street space permit         issued   
9                                       street space         issued   

  Current Status Date Filed Date Issued Date Completed Date  \
0          2017-12-21 2015-05-06  11/09/2015            NaN   
1          2017-08-03 2016-04-19  08/03/2017            NaN   
2          2017-09-26 2016-05-27         NaN            NaN   
3          2017-07-24 2016-11-07  07/18/2017     07/24/2017   
4          2017-12-01 2016-11-28  12/01/2017            NaN   
5          2017-07-06 2017-06-14  07/06/2017            NaN   
6          2017-07-12 2017-06-30  06/30/2017     07/12/2017   
7          2017-06-30 2017-06-30  06/30/2017            NaN   
8          2017-07-05 2017-07-05  07/05/2017            NaN   
9          2017-07-06 2017-07-06  07/06/2017            NaN   

  First Construction Document Date Structural Notification  \
0                       11/09/2015                     NaN   
1                       08/03/2017                     NaN   
2                              NaN                     NaN   
3                       07/18/2017                     NaN   
4                       11/20/2017                     NaN   
5                       07/06/2017                     NaN   
6                    

In [14]:
data_types = data.dtypes
data_types = pd.DataFrame(data_types)
data_types = data_types.rename(columns = {0:'ReadingDataType'})
data_types.index.names = ['FieldName']
data_types

ReadingDataType
FieldName                                             
Permit Number                                   object
Permit Type                                      int64
Permit Type Definition                          object
Permit Creation Date                    datetime64[ns]
Block                                           object
Lot                                             object
Street Number                                    int64
Street Number Suffix                            object
Street Name                                     object
Street Suffix                                   object
Unit                                           float64
Unit Suffix                                     object
Description                                     object
Current Status                                  object
Current Status Date                     datetime64[ns]
Filed Date                              datetime64[ns]
Issued Date                                     object
Completed Date                                  object
First Construction Document Date                object
Structural Notification                         object
Number of Existing Stories                     float64
Number of Proposed Stories                     float64
Voluntary Soft-Story Retrofit                   object
Fire Only Permit                                object
Permit Expiration Date                          object
Estimated Cost                                 float64
Revised Cost                                   float64
Existing Use                                    object
Existing Units                                 float64
Proposed Use                                    object
Proposed Units                                 float64
Plansets                                       float64
TIDF Compliance                                 object
Existing Construction Type                     float64
Existing Construction Type Description          object
Proposed Construction Type                     float64
Proposed Construction Type Description          object
Site Permit                                     object
Supervisor District                            float64
Neighborhoods - Analysis Boundaries             object
Zipcode                                        float64
Location                                        object
Record ID                                        int64

Ahora sí, vemos que el tipo de dato de estas columnas es `DateTime`. De hecho, deberíamos poder calcular el rango de fechas de las creaciones de permisos, por ejemplo:

In [15]:
a = data['Permit Creation Date'].max() - data['Permit Creation Date'].min()
#a / pd.Timedelta(1,'M')
a

Timedelta('2158 days 00:00:00')

In [16]:
data.describe(include='all')

Permit Number    Permit Type  Permit Type Definition  \
count         198900  198900.000000                  198900   
unique        181495            NaN                       8   
top     201602179765            NaN  otc alterations permit   
freq             101            NaN                  178844   
first            NaN            NaN                     NaN   
last             NaN            NaN                     NaN   
mean             NaN       7.522323                     NaN   
std              NaN       1.457451                     NaN   
min              NaN       1.000000                     NaN   
25%              NaN       8.000000                     NaN   
50%              NaN       8.000000                     NaN   
75%              NaN       8.000000                     NaN   
max              NaN       8.000000                     NaN   

       Permit Creation Date   Block     Lot  Street Number  \
count                198900  198900  198900  198900.000000   
unique                 1291    4896    1055            NaN   
top     2017-09-15 00:00:00    3708     001            NaN   
freq                    413    1195   10114            NaN   
first   2012-03-28 00:00:00     NaN     NaN            NaN   
last    2018-02-23 00:00:00     NaN     NaN            NaN   
mean                    NaN     NaN     NaN    1121.728944   
std                     NaN     NaN     NaN    1135.768948   
min                     NaN     NaN     NaN       0.000000   
25%                     NaN     NaN     NaN     235.000000   
50%                     NaN     NaN     NaN     710.000000   
75%                     NaN     NaN     NaN    1700.000000   
max                     NaN     NaN     NaN    8400.000000   

       Street Number Suffix Street Name Street Suffix          Unit  \
count                  2216      198900        196132  29479.000000   
unique                   18        1704            21           NaN   
top                       A      Market            St           NaN   
freq                   1501        5443        138358           NaN   
first                   NaN         NaN           NaN           NaN   
last                    NaN         NaN           NaN           NaN   
mean                    NaN         NaN           NaN     78.517182   
std                     NaN         NaN           NaN    326.981324   
min                     NaN         NaN           NaN      0.000000   
25%                     NaN         NaN           NaN      0.000000   
50%                     NaN         NaN           NaN      0.000000   
75%                     NaN         NaN           NaN      1.000000   
max                     NaN         NaN           NaN   6004.000000   

       Unit Suffix   Description Current Status  Current Status Date  \
count         1961        198610         198900               198900   
unique         164        134272             14                 1307   
top              A  street space       complete  2017-09-15 00:00:00   
freq           346         22163          97077                  474   
first          NaN           NaN            NaN  2013-01-02 00:00:00   
last           NaN           NaN            NaN  2019-03-02 00:00:00   
mean           NaN           NaN            NaN                  NaN   
std            NaN           NaN            NaN                  NaN   
min            NaN           NaN            NaN                  NaN   
25%            NaN           NaN            NaN                  NaN   
50%            NaN           NaN            NaN                  NaN   
75%            NaN           NaN            NaN                  NaN   
max            NaN           NaN            NaN                  NaN   

                 Filed Date Issued Date Completed Date  \
count                198900      183960          97191   
unique                 1288        1289           1300   
top     2017-09-15 00:00:00  06/15/2016     10/30/2015   
freq               

Con esto, podemos decir que la importación se ha realizado correctamente. Podemos pasar al punto dos de la lista.

### 1.2. Asegurar la Existencia de IDs o Claves Únicas

El siguiente paso implica chequear que no existen datos duplicados y que las claves, si existen, son únicas.

En este sentido, se podría esperar que el `Permit Number` sea único. Sin embargo, del resumen previo se puede deducir que esto no es así, ya que la cuenta es bastante superior a la cantidad de valores únicos. Es decir, existen más de 17000 permisos con el mismo `Permit Number`.

In [204]:
data.shape

(198900, 43)

In [205]:
data_ind = data.set_index("Permit Number")

In [206]:
data_ind[data_ind.index.duplicated(keep=False)].sort_index()

Permit Type  Permit Type Definition Permit Creation Date Block  \
Permit Number                                                                   
201301027072             8  otc alterations permit           2013-01-02  3582   
201301027072             8  otc alterations permit           2013-01-02  3582   
201301027079             8  otc alterations permit           2013-01-02  1230   
201301027079             8  otc alterations permit           2013-01-02  1230   
201301027081             8  otc alterations permit           2013-01-02  3196   
201301027081             8  otc alterations permit           2013-01-02  3196   
201301027093             8  otc alterations permit           2013-01-02  1635   
201301027093             8  otc alterations permit           2013-01-02  1635   
201301027095             8  otc alterations permit           2013-01-02  3564   
201301027095             8  otc alterations permit           2013-01-02  3564   
201301027105             8  otc alterations permit           2013-01-02  1173   
201301027105             8  otc alterations permit           2013-01-02  1173   
201301037165             8  otc alterations permit           2013-01-03  0535   
201301037165             8  otc alterations permit           2013-01-03  0535   
201301037171             8  otc alterations permit           2013-01-03  0520   
201301037171             8  otc alterations permit           2013-01-03  0520   
201301037188             8  otc alterations permit           2013-01-03  3643   
201301037188             8  otc alterations permit           2013-01-03  3643   
201301037191             8  otc alterations permit           2013-01-03  0998   
201301037191             8  otc alterations permit           2013-01-03  0998   
201301037195             8  otc alterations permit           2013-01-03  0976   
201301037195             8  otc alterations permit           2013-01-03  0976   
201301037195             8  otc alterations permit           2013-01-03  0976   
201301037195             8  otc alterations permit           2013-01-03  0976   
201301037207             8  otc alterations permit           2013-01-03  3732   
201301037207             8  otc alterations permit           2013-01-03  3732   
201301047213             8  otc alterations permit           2013-01-04  0849   
201301047213             8  otc alterations permit           2013-01-04  0849   
201301047218             8  otc alterations permit           2013-01-04  3537   
201301047218             8  otc alterations permit           2013-01-04  3537   
...                    ...                     ...                  ...   ...   
M888788                  8  otc alterations permit           2018-02-13  5514   
M888788                  8  otc alterations permit           2018-02-13  5514   
M888807                  8  otc alterations permit           2018-02-13  4003   
M888807                  8  otc alterations permit           2018-02-13  4003   
M889507                  8  otc alterations permit           2018-02-14  0152   
M889507                  8  otc alterations permit           2018-02-14  0152   
M889749                  8  otc alterations permit           2018-02-15  8725   
M889749                  8  otc alterations permit           2018-02-15  8725   
M889887                  8  otc alterations permit           2018-02-15  0545   
M889887                  8  otc alterations permit           2018-02-15  0545   
M889947                  8  otc alterations permit           2018-02-15  1350   
M889947                  8  otc alterations permit           2018-02-15  1350   
M890327                  8  otc alterations permit           2018-02-16  0474   
M890327                  8  otc alterations permit           2018-02-16  0474   
M890607                  8  otc alterations permit           2018-02-16  3533   
M890607                  8  otc alterations permit           2018-02-16  3533   
M890987                  8  otc alterations pe

Vemos que, al menos en los registros visualizados, la diferencia parece estar principalmente en la columna `Record ID`. Por otra parte, en las explicaciones del dataset propuesto por el autor describe este campo como "algún ID, no útil para esto". Dado que no agrega información, podemos suprimir esta columna de nuestro set de datos:

In [207]:
data_ind.columns

data_ind = data_ind[['Permit Type', 'Permit Type Definition', 'Permit Creation Date',
       'Block', 'Lot', 'Street Number', 'Street Number Suffix', 'Street Name',
       'Street Suffix', 'Unit', 'Unit Suffix', 'Description', 'Current Status',
       'Current Status Date', 'Filed Date', 'Issued Date', 'Completed Date',
       'First Construction Document Date', 'Structural Notification',
       'Number of Existing Stories', 'Number of Proposed Stories',
       'Voluntary Soft-Story Retrofit', 'Fire Only Permit',
       'Permit Expiration Date', 'Estimated Cost', 'Revised Cost',
       'Existing Use', 'Existing Units', 'Proposed Use', 'Proposed Units',
       'Plansets', 'TIDF Compliance', 'Existing Construction Type',
       'Existing Construction Type Description', 'Proposed Construction Type',
       'Proposed Construction Type Description', 'Site Permit',
       'Supervisor District', 'Neighborhoods - Analysis Boundaries', 'Zipcode',
       'Location']]
data_ind.shape

(198900, 41)

Si borramos los duplicados, considerando los valores de todas las columnas restantes, tendremos lo siguiente:

In [208]:
data_ind_uni = data_ind.drop_duplicates(subset = ['Permit Type', 'Permit Type Definition',
        'Permit Creation Date','Block', 'Lot', 'Street Number', 'Street Number Suffix', 'Street Name',
       'Street Suffix', 'Unit', 'Unit Suffix', 'Description', 'Current Status',
       'Current Status Date', 'Filed Date', 'Issued Date', 'Completed Date',
       'First Construction Document Date', 'Structural Notification',
       'Number of Existing Stories', 'Number of Proposed Stories',
       'Voluntary Soft-Story Retrofit', 'Fire Only Permit',
       'Permit Expiration Date', 'Estimated Cost', 'Revised Cost',
       'Existing Use', 'Existing Units', 'Proposed Use', 'Proposed Units',
       'Plansets', 'TIDF Compliance', 'Existing Construction Type',
       'Existing Construction Type Description', 'Proposed Construction Type',
       'Proposed Construction Type Description', 'Site Permit',
       'Supervisor District', 'Neighborhoods - Analysis Boundaries', 'Zipcode',
       'Location'], inplace=False)

data_ind_uni.shape

(198627, 41)

Vemos que sólo se han eliminado 273 registros, por lo que aún permanecen `Permit Numbers` repetidos. Analicemos nuevamente los valores de los restantes campos:

In [209]:
data_ind_dupl = data_ind_uni[data_ind_uni.index.duplicated(keep=False)].sort_index()
# data_ind_dupl.to_csv("Output/analisis_permit_num_repetido.csv")
data_ind_dupl

Permit Type  Permit Type Definition Permit Creation Date Block  \
Permit Number                                                                   
201301027072             8  otc alterations permit           2013-01-02  3582   
201301027072             8  otc alterations permit           2013-01-02  3582   
201301027079             8  otc alterations permit           2013-01-02  1230   
201301027079             8  otc alterations permit           2013-01-02  1230   
201301027081             8  otc alterations permit           2013-01-02  3196   
201301027081             8  otc alterations permit           2013-01-02  3196   
201301027093             8  otc alterations permit           2013-01-02  1635   
201301027093             8  otc alterations permit           2013-01-02  1635   
201301027095             8  otc alterations permit           2013-01-02  3564   
201301027095             8  otc alterations permit           2013-01-02  3564   
201301027105             8  otc alterations permit           2013-01-02  1173   
201301027105             8  otc alterations permit           2013-01-02  1173   
201301037165             8  otc alterations permit           2013-01-03  0535   
201301037165             8  otc alterations permit           2013-01-03  0535   
201301037171             8  otc alterations permit           2013-01-03  0520   
201301037171             8  otc alterations permit           2013-01-03  0520   
201301037188             8  otc alterations permit           2013-01-03  3643   
201301037188             8  otc alterations permit           2013-01-03  3643   
201301037191             8  otc alterations permit           2013-01-03  0998   
201301037191             8  otc alterations permit           2013-01-03  0998   
201301037195             8  otc alterations permit           2013-01-03  0976   
201301037195             8  otc alterations permit           2013-01-03  0976   
201301037195             8  otc alterations permit           2013-01-03  0976   
201301037195             8  otc alterations permit           2013-01-03  0976   
201301037207             8  otc alterations permit           2013-01-03  3732   
201301037207             8  otc alterations permit           2013-01-03  3732   
201301047213             8  otc alterations permit           2013-01-04  0849   
201301047213             8  otc alterations permit           2013-01-04  0849   
201301047218             8  otc alterations permit           2013-01-04  3537   
201301047218             8  otc alterations permit           2013-01-04  3537   
...                    ...                     ...                  ...   ...   
M888788                  8  otc alterations permit           2018-02-13  5514   
M888788                  8  otc alterations permit           2018-02-13  5514   
M888807                  8  otc alterations permit           2018-02-13  4003   
M888807                  8  otc alterations permit           2018-02-13  4003   
M889507                  8  otc alterations permit           2018-02-14  0152   
M889507                  8  otc alterations permit           2018-02-14  0152   
M889749                  8  otc alterations permit           2018-02-15  8725   
M889749                  8  otc alterations permit           2018-02-15  8725   
M889887                  8  otc alterations permit           2018-02-15  0545   
M889887                  8  otc alterations permit           2018-02-15  0545   
M889947                  8  otc alterations permit           2018-02-15  1350   
M889947                  8  otc alterations permit           2018-02-15  1350   
M890327                  8  otc alterations permit           2018-02-16  0474   
M890327                  8  otc alterations permit           2018-02-16  0474   
M890607                  8  otc alterations permit           2018-02-16  3533   
M890607                  8  otc alterations permit           2018-02-16  3533   
M890987                  8  otc alterations pe

En muchos casos, se observa que tanto `Street Number` como `Lot` pueden tener diferentes valores para el mismo `Permit Number`. Esto es algo factible, ya que el mismo permiso pudo haber aplicado a diferentes construcciones. Construimos una clave que los incorpore:

In [210]:
data_ind_uni.reset_index(inplace = True)

In [211]:
data_ind_uni['ID'] = data_ind_uni['Permit Number'] + ' - ' + data_ind_uni['Lot'] + ' - ' + data_ind_uni['Street Number'].astype(str)
data_ind_uni.head(2)

C:\Users\Julieta\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


Permit Number  Permit Type Permit Type Definition Permit Creation Date  \
0  201505065519            4           sign - erect           2015-05-06   
1  201604195146            4           sign - erect           2016-04-19   

  Block  Lot  Street Number Street Number Suffix Street Name Street Suffix  \
0  0326  023            140                  NaN       Ellis            St   
1  0306  007            440                  NaN       Geary            St   

   Unit Unit Suffix                                        Description  \
0   NaN         NaN  ground fl facade: to erect illuminated, electr...   
1   0.0         NaN            remove (e) awning and associated signs.   

  Current Status Current Status Date Filed Date Issued Date Completed Date  \
0        expired          2017-12-21 2015-05-06  11/09/2015            NaN   
1         issued          2017-08-03 2016-04-19  08/03/2017            NaN   

  First Construction Document Date Structural Notification  \
0                       11/09/2015                     NaN   
1                       08/03/2017                     NaN   

   Number of Existing Stories  Number of Proposed Stories  \
0                         6.0                         NaN   
1                         7.0                         NaN   

  Voluntary Soft-Story Retrofit Fire Only Permit Permit Expiration Date  \
0                           NaN              NaN             11/03/2016   
1                           NaN              NaN             12/03/2017   

   Estimated Cost  Revised Cost         Existing Use  Existing Units  \
0          4000.0        4000.0  tourist hotel/motel           143.0   
1             1.0         500.0  tourist hotel/motel             NaN   

  Proposed Use  Proposed Units  Plansets TIDF Compliance  \
0          NaN             NaN       2.0             NaN   
1          NaN             NaN       2.0             NaN   

   Existing Construction Type Existing Construction Type Description  \
0                         3.0                          constr type 3   
1                         3.0                          constr type 3   

   Proposed Construction Type Proposed Construction Type Description  \
0                         NaN                                    NaN   
1                         NaN                                    NaN   

  Site Permit  Supervisor District Neighborhoods - Analysis Boundaries  \
0         NaN                  3.0                          Tenderloin   
1         NaN                  3.0                          Tenderloin   

   Zipcode                                   Location  \
0  94102.0  (37.785719256680785, -122.40852313194863)   
1  94102.0   (37.78733980600732, -122.41063199757738)   

                         ID  
0  201505065519 - 023 - 140  
1  201604195146 - 007 - 440

Una vez que obtenemos este ID, volvemos a preguntar por los valores duplicados:

In [212]:
data_ind_uni.set_index('ID', inplace = True)

In [213]:
data_ind_dupl = data_ind_uni[data_ind_uni.index.duplicated(keep=False)].sort_index()

data_ind_dupl.head(20)

Permit Number  Permit Type  \
ID                                                     
201302049435 - 062 - 771   201302049435            8   
201302049435 - 062 - 771   201302049435            8   
201302059475 - 098 - 511   201302059475            8   
201302059475 - 098 - 511   201302059475            8   
201302089862 - 009A - 234  201302089862            8   
201302089862 - 009A - 234  201302089862            8   
201302119974 - 016 - 436   201302119974            8   
201302119974 - 016 - 436   201302119974            8   
201303081867 - 016 - 2592  201303081867            8   
201303081867 - 016 - 2592  201303081867            8   
201303132112 - 040 - 194   201303132112            8   
201303132112 - 040 - 194   201303132112            8   
201303132112 - 040 - 196   201303132112            8   
201303132112 - 040 - 196   201303132112            8   
201303152346 - 020 - 1223  201303152346            8   
201303152346 - 020 - 1223  201303152346            8   
201303202663 - 014 - 2878  201303202663            8   
201303202663 - 014 - 2878  201303202663            8   
201304023644 - 047A - 62   201304023644            3   
201304023644 - 047A - 62   201304023644            3   

                                     Permit Type Definition  \
ID                                                            
201302049435 - 062 - 771             otc alterations permit   
201302049435 - 062 - 771             otc alterations permit   
201302059475 - 098 - 511             otc alterations permit   
201302059475 - 098 - 511             otc alterations permit   
201302089862 - 009A - 234            otc alterations permit   
201302089862 - 009A - 234            otc alterations permit   
201302119974 - 016 - 436             otc alterations permit   
201302119974 - 016 - 436             otc alterations permit   
201303081867 - 016 - 2592            otc alterations permit   
201303081867 - 016 - 2592            otc alterations permit   
201303132112 - 040 - 194             otc alterations permit   
201303132112 - 040 - 194             otc alterations permit   
201303132112 - 040 - 196             otc alterations permit   
201303132112 - 040 - 196             otc alterations permit   
201303152346 - 020 - 1223            otc alterations permit   
201303152346 - 020 - 1223            otc alterations permit   
201303202663 - 014 - 2878            otc alterations permit   
201303202663 - 014 - 2878            otc alterations permit   
201304023644 - 047A - 62   additions alterations or repairs   
201304023644 - 047A - 62   additions alterations or repairs   

                          Permit Creation Date Block   Lot  Street Number  \
ID                                                                          
201302049435 - 062 - 771            2013-02-04  3729   062            771   
201302049435 - 062 - 771            2013-02-04  3729   062            771   
201302059475 - 098 - 511            2013-02-05  3726   098            511   
201302059475 - 098 - 511            2013-02-05  3726   098            511   
201302089862 - 009A - 234           2013-02-08  0868  009A            234   
201302089862 - 009A - 234           2013-02-08  0868  009A            234   
201302119974 - 016 - 436            2013-02-11  1020   016            436   
201302119974 - 016 - 436            2013-02-11  1020   016            436   
201303081867 - 016 - 2592           2013-03-08  1053   016           2592   
201303081867 - 016 - 2592           2013-03-08  1053   016           2592   
201303132112 - 040 - 194            2013-03-13  6615   040            194   
201303132112 - 040 - 194            2013-03-13  6615   040            194   
201303132112 - 040 - 196            2013-03-13  6615   040            196   
201303132112 - 040 - 196            2013-03-13  6615   040            196   
201303152346 - 020 - 1223           2013-03-15  5654   020           1223   
201303152346 - 020 - 1223           2013-03-15  5654   020           1223   
201303202663 - 014 -

In [214]:
data_ind_dupl.shape

(893, 42)

De 31489 `Permit Numbers` duplicados, pasamos a tan solo 893 registros repetidos. Además, vemos que la principal diferencia de valores entre `ID's` repetidos se produce en las columnas `Street Number Suffix`, `Street Name`, `Street Suffix`, `Unit` y `Unit Suffix`.En el caso del nombre de la calle y sufijo, probablemente se trate de construcciones en esquinas de manzana, que dan a dos calles. Pero como ya existe el dato de localización con coordenadas, es indistinto que mantengamos una calle u otra.

Para asegurarnos que la eliminación de duplicados tome aquellos registros que no tienen NaN en estos campos, ordenamos inicialmente el dataset, de acuerdo a la fecha del estado actual (`Current Status Date`), de modo de mantener la fecha más reciente, ya que será la última versión del registro, y dejando los valores NaN al final.

In [309]:
data_ind_uni.sort_values(by = ['Current Status Date'], axis=0, ascending=False, inplace=True,
                         na_position='last')

data_ind_uni.tail(5)

C:\Users\Julieta\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


Permit Number  Permit Type  Permit Type Definition  \
ID                                                                            
M364307 - 045 - 2300           M364307            8  otc alterations permit   
M364388 - 026 - 2615           M364388            8  otc alterations permit   
M364287 - 003 - 105            M364287            8  otc alterations permit   
201301027082 - 018 - 210  201301027082            8  otc alterations permit   
M364389 - 031A - 30            M364389            8  otc alterations permit   

                         Permit Creation Date Block   Lot  Street Number  \
ID                                                                         
M364307 - 045 - 2300               2013-01-02  2362   045           2300   
M364388 - 026 - 2615               2013-01-02  0951   026           2615   
M364287 - 003 - 105                2013-01-02  0493   003            105   
201301027082 - 018 - 210           2013-01-02  5895   018            210   
M364389 - 031A - 30                2013-01-02  3555  031A             30   

                         Street Number Suffix Street Name Street Suffix  Unit  \
ID                                                                              
M364307 - 045 - 2300                      NaN     Taraval            St   NaN   
M364388 - 026 - 2615                      NaN       Union            St   NaN   
M364287 - 003 - 105                       NaN    Magnolia            St   NaN   
201301027082 - 018 - 210                  NaN     Maynard            St   NaN   
M364389 - 031A - 30                       NaN      Albion            St   NaN   

                         Unit Suffix  \
ID                                     
M364307 - 045 - 2300             NaN   
M364388 - 026 - 2615             NaN   
M364287 - 003 - 105              NaN   
201301027082 - 018 - 210         NaN   
M364389 - 031A - 30              NaN   

                                                                Description  \
ID                                                                            
M364307 - 045 - 2300                                           street space   
M364388 - 026 - 2615                                           street space   
M364287 - 003 - 105                                     street space permit   
201301027082 - 018 - 210  remove extg cabinets, lathe, & plaster in kitc...   
M364389 - 031A - 30                        street space and sidewalk repair   

                         Current Status Current Status Date Filed Date  \
ID                                                                       
M364307 - 045 - 2300             issued          2013-01-02 2013-01-02   
M364388 - 026 - 2615             issued          2013-01-02 2013-01-02   
M364287 - 003 - 105              issued          2013-01-02 2013-01-02   
201301027082 - 018 - 210         issued          2013-01-02 2013-01-02   
M364389 - 031A - 30              issued          2013-01-02 2013-01-02   

                         Issued Date Completed Date  \
ID                                                    
M364307 - 045 - 2300      01/02/2013            NaN   
M364388 - 026 - 2615      01/02/2013            NaN   
M364287 - 003 - 105       01/02/2013            NaN   
201301027082 - 018 - 210  01/02/2013            NaN   
M364389 - 031A - 30       01/02/2013            NaN   

                         First Construction Document Date  \
ID                                                          
M364307 - 045 - 2300                           01/02/2013   
M364388 - 026 - 2615                           01/02/2013   
M364287 - 003 - 105                            01/02/2013   
201301027082 - 018 - 210                       01/02/2013   
M364389 - 031A - 30                            01/02/2013   

                         Structural Notification  Number of Existing Stories  \
ID                                                                             
M364307 - 045 - 2300                      

In [348]:
data_ind_uni2 = data_ind_uni.drop_duplicates(subset = ['Permit Number','Lot', 'Street Number'], keep = 'first')

data_ind_uni2.shape

(198138, 42)

In [349]:
data_ind_uni2.head()

Permit Number  Permit Type  Permit Type Definition  \
ID                                                                              
201611293622 - 005D - 1391  201611293622            8  otc alterations permit   
201802161497 - 003 - 3400   201802161497            8  otc alterations permit   
201710232023 - 016 - 682    201710232023            8  otc alterations permit   
201709209051 - 007A - 1980  201709209051            8  otc alterations permit   
201802060459 - 029 - 345    201802060459            4            sign - erect   

                           Permit Creation Date Block   Lot  Street Number  \
ID                                                                           
201611293622 - 005D - 1391           2016-11-29  0480  005D           1391   
201802161497 - 003 - 3400            2018-02-16  0471   003           3400   
201710232023 - 016 - 682             2017-10-23  0828   016            682   
201709209051 - 007A - 1980           2017-09-20  0624  007A           1980   
201802060459 - 029 - 345             2018-02-06  5598   029            345   

                           Street Number Suffix Street Name Street Suffix  \
ID                                                                          
201611293622 - 005D - 1391                  NaN   Francisco            St   
201802161497 - 003 - 3400                   NaN      Laguna            St   
201710232023 - 016 - 682                    NaN         Oak            St   
201709209051 - 007A - 1980                  NaN  Sacramento            St   
201802060459 - 029 - 345                    NaN   Bay Shore            Bl   

                            Unit Unit Suffix  \
ID                                             
201611293622 - 005D - 1391   NaN         NaN   
201802161497 - 003 - 3400    NaN         NaN   
201710232023 - 016 - 682     NaN         NaN   
201709209051 - 007A - 1980   NaN         NaN   
201802060459 - 029 - 345     NaN         NaN   

                                                                  Description  \
ID                                                                              
201611293622 - 005D - 1391  kitchen: reface cabinets, new sinks, faucet an...   
201802161497 - 003 - 3400   unit #227- cosmetic renovation of living room ...   
201710232023 - 016 - 682    to obtain a final inspection for work approved...   
201709209051 - 007A - 1980  soft story retrofit per sfebc chapter 4d, 2016...   
201802060459 - 029 - 345    install 1 non-illuminated single faced clearan...   

                           Current Status Current Status Date Filed Date  \
ID                                                                         
201611293622 - 005D - 1391      cancelled          2019-03-02 2016-11-29   
201802161497 - 003 - 3400          issued          2018-02-23 2018-02-16   
201710232023 - 016 - 682         complete          2018-02-23 2017-10-23   
201709209051 - 007A - 1980         issued          2018-02-23 2017-09-20   
201802060459 - 029 - 345           issued          2018-02-23 2018-02-06   

                           Issued Date Completed Date  \
ID                                                      
201611293622 - 005D - 1391  11/29/2016            NaN   
201802161497 - 003 - 3400   02/23/2018            NaN   
201710232023 - 016 - 682    10/23/2017     02/23/2018   
201709209051 - 007A - 1980  02/23/2018            NaN   
201802060459 - 029 - 345    02/23/2018            NaN   

                           First Construction Document Date  \
ID                                                            
201611293622 - 005D - 1391                       11/29/2016   
201802161497 - 003 - 3400                        02/23/2018   
201710232023 - 016 - 682                         10/23/2017   
201709209051 - 007A - 1980                       02/23/2018   
201802060459 - 029 - 345                         02/23/2018   

                           Structural Notification  \
ID                                    

Se eliminaron 489 registros más. Repetimos el procedimiento para volver a analizar los duplicados remanentes.

In [350]:
data_ind_dupl = data_ind_uni2[data_ind_uni2.index.duplicated(keep=False)].sort_index()
data_ind_dupl.shape

(0, 42)

In [351]:
data_ind_dupl

Empty DataFrame
Columns: [Permit Number, Permit Type, Permit Type Definition, Permit Creation Date, Block, Lot, Street Number, Street Number Suffix, Street Name, Street Suffix, Unit, Unit Suffix, Description, Current Status, Current Status Date, Filed Date, Issued Date, Completed Date, First Construction Document Date, Structural Notification, Number of Existing Stories, Number of Proposed Stories, Voluntary Soft-Story Retrofit, Fire Only Permit, Permit Expiration Date, Estimated Cost, Revised Cost, Existing Use, Existing Units, Proposed Use, Proposed Units, Plansets, TIDF Compliance, Existing Construction Type, Existing Construction Type Description, Proposed Construction Type, Proposed Construction Type Description, Site Permit, Supervisor District, Neighborhoods - Analysis Boundaries, Zipcode, Location]
Index: []

Ahora sí, logramos un registro único por cada `Permit Number`, `Lot` y `Street Number`.

In [352]:
data_unique = data_ind_uni2

Luego de explorar el dataset, además de las fechas tipificadas en la metadata, encontramos nuevos campos que contienen fechas: `Issued Date`, `First Construction Document Date` y `Permit Expiration Date`. Vamos a redefinir estos tipos de datos.

In [353]:
data_unique.columns

data_unique['Issued Date'] = pd.to_datetime(data_unique['Issued Date'])
data_unique['First Construction Document Date'] = pd.to_datetime(data_unique['First Construction Document Date'])
data_unique['Permit Expiration Date'] = pd.to_datetime(data_unique['Permit Expiration Date'])

C:\Users\Julieta\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  This is separate from the ipykernel package so we can avoid doing imports until
C:\Users\Julieta\Anaconda3\lib\site-packages\ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  after removing the cwd from sys.path.
C:\Users\Julieta\Anaconda3\lib\site-packages\ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_ind

In [355]:
data_unique.head(10)

Permit Number  Permit Type  Permit Type Definition  \
ID                                                                              
201611293622 - 005D - 1391  201611293622            8  otc alterations permit   
201802161497 - 003 - 3400   201802161497            8  otc alterations permit   
201710232023 - 016 - 682    201710232023            8  otc alterations permit   
201709209051 - 007A - 1980  201709209051            8  otc alterations permit   
201802060459 - 029 - 345    201802060459            4            sign - erect   
201802060457 - 029 - 345    201802060457            4            sign - erect   
201802060455 - 029 - 345    201802060455            4            sign - erect   
201802060451 - 029 - 345    201802060451            4            sign - erect   
201802221983 - 062 - 303    201802221983            8  otc alterations permit   
201801310068 - 040 - 50     201801310068            8  otc alterations permit   

                           Permit Creation Date Block   Lot  Street Number  \
ID                                                                           
201611293622 - 005D - 1391           2016-11-29  0480  005D           1391   
201802161497 - 003 - 3400            2018-02-16  0471   003           3400   
201710232023 - 016 - 682             2017-10-23  0828   016            682   
201709209051 - 007A - 1980           2017-09-20  0624  007A           1980   
201802060459 - 029 - 345             2018-02-06  5598   029            345   
201802060457 - 029 - 345             2018-02-06  5598   029            345   
201802060455 - 029 - 345             2018-02-06  5598   029            345   
201802060451 - 029 - 345             2018-02-06  5598   029            345   
201802221983 - 062 - 303             2018-02-22  3749   062            303   
201801310068 - 040 - 50              2018-01-31  1085   040             50   

                           Street Number Suffix Street Name Street Suffix  \
ID                                                                          
201611293622 - 005D - 1391                  NaN   Francisco            St   
201802161497 - 003 - 3400                   NaN      Laguna            St   
201710232023 - 016 - 682                    NaN         Oak            St   
201709209051 - 007A - 1980                  NaN  Sacramento            St   
201802060459 - 029 - 345                    NaN   Bay Shore            Bl   
201802060457 - 029 - 345                    NaN   Bay Shore            Bl   
201802060455 - 029 - 345                    NaN   Bay Shore            Bl   
201802060451 - 029 - 345                    NaN   Bay Shore            Bl   
201802221983 - 062 - 303                    NaN        02nd            St   
201801310068 - 040 - 50                     NaN     Stanyan            St   

                            Unit Unit Suffix  \
ID                                             
201611293622 - 005D - 1391   NaN         NaN   
201802161497 - 003 - 3400    NaN         NaN   
201710232023 - 016 - 682     NaN         NaN   
201709209051 - 007A - 1980   NaN         NaN   
201802060459 - 029 - 345     NaN         NaN   
201802060457 - 029 - 345     NaN         NaN   
201802060455 - 029 - 345     NaN         NaN   
201802060451 - 029 - 345     NaN         NaN   
201802221983 - 062 - 303     NaN         NaN   
201801310068 - 040 - 50      0.0         NaN   

                                                                  Description  \
ID                                                                              
201611293622 - 005D - 1391  kitchen: reface cabinets, new sinks, faucet an...   
201802161497 - 003 - 3400   unit #227- cosmetic renovation of living room ...   
201710232023 - 016 - 682    to obtain a final inspection for work approved...   
201709209051 - 007A - 1980  soft story retrofit per sfebc chapter 4d, 2016...   
201802060459 - 029 - 345    install 1 non-illuminated single faced clearan...   
201802060457 - 029 - 345    install 3 illulminat

### 1.3. Despersonalizar Datos y Guardarlos en un Nuevo Archivo

En este caso, no es necesario despersonalizar ningún dato, dado que no existe información personal sensible.

### 1.4. Nunca Modificar los Datos Crudos u Originales

Al finalizar la limpieza, guardaremos nuevamente el dataset, para asegurarnos de no modificar los datos originales.

## 2. Pasos de Limpieza Necesarios

### 2.1. Etiquetas de Variables/Columnas y Problemas de Codificación/Encoding

Antes que nada, comprobamos el encoding de la fuente de datos, leyendo en crudo los primeros 100000 caracteres:

In [316]:
import chardet

In [317]:
# with open("Building_Permits.csv", 'rb') as rawdata:
#     result = chardet.detect(rawdata.read(100000))
# result

import requests
rawdata = requests.get("https://github.com/sac2087/famaf/blob/master/Analisis%20exploratorio%20y%20curacion%20de%20datos/Ejercicios/Building_Permits.csv?raw=true")
result = chardet.detect(rawdata.content[:100000])
result

{'confidence': 1.0, 'encoding': 'ascii', 'language': ''}

El set de datos está convenientemente codificado en ASCII, por lo que no tendremos problemas al importar.

Realizamos ahora un resumen descriptivo del dataset con valores únicos:

In [319]:
data_unique.reset_index().describe(include = 'all')

ID Permit Number    Permit Type  \
count                     198138        198138  198138.000000   
unique                    198138        181298            NaN   
top     201606018835 - 025 - 300  201602179765            NaN   
freq                           1            68            NaN   
first                        NaN           NaN            NaN   
last                         NaN           NaN            NaN   
mean                         NaN           NaN       7.525543   
std                          NaN           NaN       1.453203   
min                          NaN           NaN       1.000000   
25%                          NaN           NaN       8.000000   
50%                          NaN           NaN       8.000000   
75%                          NaN           NaN       8.000000   
max                          NaN           NaN       8.000000   

        Permit Type Definition Permit Creation Date   Block     Lot  \
count                   198138               198138  198138  198138   
unique                       8                 1291    4896    1055   
top     otc alterations permit  2017-09-15 00:00:00    3708     001   
freq                    178299                  409    1194   10065   
first                      NaN  2012-03-28 00:00:00     NaN     NaN   
last                       NaN  2018-02-23 00:00:00     NaN     NaN   
mean                       NaN                  NaN     NaN     NaN   
std                        NaN                  NaN     NaN     NaN   
min                        NaN                  NaN     NaN     NaN   
25%                        NaN                  NaN     NaN     NaN   
50%                        NaN                  NaN     NaN     NaN   
75%                        NaN                  NaN     NaN     NaN   
max                        NaN                  NaN     NaN     NaN   

        Street Number Street Number Suffix Street Name Street Suffix  \
count   198138.000000                 1941      198138        195371   
unique            NaN                   17        1704            21   
top               NaN                    A      Market            St   
freq              NaN                 1342        5430        137859   
first             NaN                  NaN         NaN           NaN   
last              NaN                  NaN         NaN           NaN   
mean      1122.051585                  NaN         NaN           NaN   
std       1136.082463                  NaN         NaN           NaN   
min          0.000000                  NaN         NaN           NaN   
25%        235.000000                  NaN         NaN           NaN   
50%        710.000000                  NaN         NaN           NaN   
75%       1700.000000                  NaN         NaN           NaN   
max       8400.000000                  NaN         NaN           NaN   

                Unit Unit Suffix   Description Current Status  \
count   29396.000000        1924        197855         198138   
unique           NaN         158        134266             14   
top              NaN           A  street space       complete   
freq             NaN         338         22071          96787   
first            NaN         NaN           NaN            NaN   
last             NaN         NaN           NaN            NaN   
mean       78.692645         NaN           NaN            NaN   
std       327.385628         NaN           NaN            NaN   
min         0.000000         NaN           NaN            NaN   
25%         0.000000         NaN           NaN            NaN   
50%         0.000000         NaN           NaN            NaN   
75%         1.000000         NaN           NaN            NaN   
max      6004.000000         NaN           NaN            NaN   

        Current Status Date           Filed Date          Issued Date  \
count                198138               198138               183328   
unique                 1307                 1288          

Analizamos los nombres de columnas:

In [356]:
data_unique.columns

Index(['Permit Number', 'Permit Type', 'Permit Type Definition',
       'Permit Creation Date', 'Block', 'Lot', 'Street Number',
       'Street Number Suffix', 'Street Name', 'Street Suffix', 'Unit',
       'Unit Suffix', 'Description', 'Current Status', 'Current Status Date',
       'Filed Date', 'Issued Date', 'Completed Date',
       'First Construction Document Date', 'Structural Notification',
       'Number of Existing Stories', 'Number of Proposed Stories',
       'Voluntary Soft-Story Retrofit', 'Fire Only Permit',
       'Permit Expiration Date', 'Estimated Cost', 'Revised Cost',
       'Existing Use', 'Existing Units', 'Proposed Use', 'Proposed Units',
       'Plansets', 'TIDF Compliance', 'Existing Construction Type',
       'Existing Construction Type Description', 'Proposed Construction Type',
       'Proposed Construction Type Description', 'Site Permit',
       'Supervisor District', 'Neighborhoods - Analysis Boundaries', 'Zipcode',
       'Location'],
      dtype='objec

In [357]:
data_unique.columns[data_unique.columns != data_unique.columns.str.extract(r'^(\w+)$')]

C:\Users\Julieta\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)
  """Entry point for launching an IPython kernel.


Index(['Permit Number', 'Permit Type', 'Permit Type Definition',
       'Permit Creation Date', 'Street Number', 'Street Number Suffix',
       'Street Name', 'Street Suffix', 'Unit Suffix', 'Current Status',
       'Current Status Date', 'Filed Date', 'Issued Date', 'Completed Date',
       'First Construction Document Date', 'Structural Notification',
       'Number of Existing Stories', 'Number of Proposed Stories',
       'Voluntary Soft-Story Retrofit', 'Fire Only Permit',
       'Permit Expiration Date', 'Estimated Cost', 'Revised Cost',
       'Existing Use', 'Existing Units', 'Proposed Use', 'Proposed Units',
       'TIDF Compliance', 'Existing Construction Type',
       'Existing Construction Type Description', 'Proposed Construction Type',
       'Proposed Construction Type Description', 'Site Permit',
       'Supervisor District', 'Neighborhoods - Analysis Boundaries'],
      dtype='object')

Vemos que no hay caracteres extraños (distintos de a-Z, 0-9 y _ ), a excepción de los nombres con espacios y "-", por lo que proponemos quitarlos:

In [358]:
new_cols = (data_unique.columns.str.replace(' ', '')).str.replace('-','')

data_unique.columns = new_cols
data_unique

PermitNumber  PermitType  \
ID                                                     
201611293622 - 005D - 1391  201611293622           8   
201802161497 - 003 - 3400   201802161497           8   
201710232023 - 016 - 682    201710232023           8   
201709209051 - 007A - 1980  201709209051           8   
201802060459 - 029 - 345    201802060459           4   
201802060457 - 029 - 345    201802060457           4   
201802060455 - 029 - 345    201802060455           4   
201802060451 - 029 - 345    201802060451           4   
201802221983 - 062 - 303    201802221983           8   
201801310068 - 040 - 50     201801310068           8   
201708175087 - 011 - 668    201708175087           6   
201708164984 - 004G - 2348  201708164984           8   
201708164984 - 004G - 2346  201708164984           8   
201708164982 - 004G - 2344  201708164982           8   
201708164982 - 004G - 2342  201708164982           8   
201802221997 - 001 - 2675   201802221997           8   
201708164983 - 004G - 2352  201708164983           8   
201708164983 - 004G - 2350  201708164983           8   
201802151475 - 015 - 2250   201802151475           8   
201708164985 - 004G - 2340  201708164985           8   
201708164985 - 004G - 2338  201708164985           8   
201708164981 - 004G - 2356  201708164981           8   
201802221963 - 013 - 2544   201802221963           8   
201801188881 - 037 - 2861   201801188881           8   
201405286908 - 062 - 1075   201405286908           1   
201801259494 - 053 - 208    201801259494           8   
201407080697 - 016 - 682    201407080697           8   
201605046574 - 031 - 462    201605046574           8   
201603010901 - 019 - 1565   201603010901           3   
201602290755 - 057 - 47     201602290755           3   
...                                  ...         ...   
201301037111 - 011 - 101    201301037111           8   
201301027081 - 011 - 1616   201301027081           8   
M364327 - 038 - 121              M364327           8   
201301027079 - 017 - 588    201301027079           8   
M364407 - 640 - 338              M364407           8   
201301027103 - 001 - 101    201301027103           8   
M364387 - 112 - 62               M364387           8   
M364347 - 025 - 988              M364347           8   
M364408 - 017 - 673              M364408           8   
M364267 - 049A - 839             M364267           8   
M364348 - 001 - 355              M364348           8   
M364367 - 072 - 69               M364367           8   
201301027090 - 042 - 865    201301027090           8   
M364368 - 007D - 2038            M364368           8   
M364229 - 019 - 350              M364229           8   
M364228 - 011A - 920             M364228           8   
M364247 - 001A - 2757            M364247           8   
201301027056 - 008 - 444    201301027056           8   
M364427 - 028 - 1235             M364427           8   
201301027081 - 011 - 1600   201301027081           8   
201301027079 - 017 - 580    201301027079           8   
201301027076 - 009 - 2335   201301027076           8   
M364227 - 026 - 3225             M364227           8   
201301027077 - 023 - 645    201301027077           8   
201301027045 - 004 - 136    201301027045           8   
M364307 - 045 - 2300             M364307           8   
M364388 - 026 - 2615             M364388           8   
M364287 - 003 - 105              M364287           8   
201301027082 - 018 - 210    201301027082           8   
M364389 - 031A - 30              M364389           8   

                                        PermitTypeDefinition  \
ID                                                             
201611293622 - 005D - 1391            otc alterations permit   
201802161497 - 003 - 3400             otc alterations permit   
201710232023 - 016 - 682              otc alterations permit   
201709209051 - 007A - 1980            otc alterations permit   
201802060459 - 029 - 345                        sign - erect   
201802060457 - 029 - 345                        sig

In [359]:
data_unique.columns[data_unique.columns != data_unique.columns.str.extract(r'^(\w+)$')]

C:\Users\Julieta\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)
  """Entry point for launching an IPython kernel.


Index([], dtype='object')

Analicemos también si hay valores con caracteres extraños dentro del dataset:

In [324]:
import ftfy.badness as bad

In [326]:
def weird(val):
    if isinstance(val, float): 
        return 0
    return bad.sequence_weirdness(val)

data_unique['DescriptionWeird'] = data_unique['Description'].apply(weird)
data_unique[data_unique['DescriptionWeird'] > 1]

C:\Users\Julieta\Anaconda3\lib\site-packages\ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  import sys


Empty DataFrame
Columns: [PermitNumber, PermitType, PermitTypeDefinition, PermitCreationDate, Block, Lot, StreetNumber, StreetNumberSuffix, StreetName, StreetSuffix, Unit, UnitSuffix, Description, CurrentStatus, CurrentStatusDate, FiledDate, IssuedDate, CompletedDate, FirstConstructionDocumentDate, StructuralNotification, NumberOfExistingStories, NumberOfProposedStories, VoluntarySoftStoryRetrofit, FireOnlyPermit, PermitExpirationDate, EstimatedCost, RevisedCost, ExistingUse, ExistingUnits, ProposedUse, ProposedUnits, Plansets, TIDFCompliance, ExistingConstructionType, ExistingConstructionTypeDescription, ProposedConstructionType, ProposedConstructionTypeDescription, SitePermit, Supervisoristrict, NeighborhoodsAnalysisBoundaries, Zipcode, Location, DescriptionWeird]
Index: []

In [327]:
data_unique['ExistingUseWeird'] = data_unique['ExistingUse'].apply(weird)
data_unique[data_unique['ExistingUseWeird'] > 1]

C:\Users\Julieta\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


Empty DataFrame
Columns: [PermitNumber, PermitType, PermitTypeDefinition, PermitCreationDate, Block, Lot, StreetNumber, StreetNumberSuffix, StreetName, StreetSuffix, Unit, UnitSuffix, Description, CurrentStatus, CurrentStatusDate, FiledDate, IssuedDate, CompletedDate, FirstConstructionDocumentDate, StructuralNotification, NumberOfExistingStories, NumberOfProposedStories, VoluntarySoftStoryRetrofit, FireOnlyPermit, PermitExpirationDate, EstimatedCost, RevisedCost, ExistingUse, ExistingUnits, ProposedUse, ProposedUnits, Plansets, TIDFCompliance, ExistingConstructionType, ExistingConstructionTypeDescription, ProposedConstructionType, ProposedConstructionTypeDescription, SitePermit, Supervisoristrict, NeighborhoodsAnalysisBoundaries, Zipcode, Location, DescriptionWeird, ExistingUseWeird]
Index: []

En los campos más controversiales, no encontramos caracteres extraños.

### 2.2. Tratamiento de Valores Faltantes

Para analizar los valores faltantes, primero veamos cuántos tenemos por campo y cuánto representan del total:

In [360]:
valores_faltantes = pd.DataFrame([data_unique.isnull().sum(),
                                   data_unique.isnull().sum()/len(data_unique)]).transpose().rename(
    columns = {0:'Cantidad_NaN',1:'Porcentaje_Nan_s_Total'})



#valores_faltantes.style.format({'Porcentaje_Nan_s_Total':"{:.2%}"})
valores_faltantes.loc[valores_faltantes['Cantidad_NaN']>0].style.format({'Porcentaje_Nan_s_Total':"{:.2%}"})

In [343]:
len(data_unique.dropna())/len(data_unique)

0.0

Si bien hay muchos valores faltantes, en varias de las columnas, vemos que en muchas de ellas seguramente es correcto que el dato sea inexistente, como por ejemplo `StreetNumberSuffix`, ya que no todos los números de calle tienen sufijos. Adicionalmente, se arriba a esta conclusion a partir de que **todos los registros tienen al menos un valor NaN**. Y por cuestiones de espacio, puede ser muy conveniente almacenar los datos de este modo en tales columnas.
En particular, en todas aquellas columnas con un alto porcentaje de vacíos debería ser común y posible que no exista dato (superior al 80%).

In [363]:
valores_faltantes.loc[(valores_faltantes['Cantidad_NaN']>0) & (valores_faltantes['Porcentaje_Nan_s_Total']<0.8)].style.format({'Porcentaje_Nan_s_Total':"{:.2%}"})

En muchos casos, falta información sobre el costo, el uso existente y el propuesto de la unidad, el tipo de las construcciones existente y propuesta y, en muy pocos casos (menos del 1%), faltan datos asociados a la localización. Otro caso llamativo es el de `CompletedDate`, sin embargo debe tratarse de permisos cuya estado aún no es `complete`.

De todos modos, en ningún caso parece pertinente descartar los datos de algún permiso o llenar estos campos con algún valor específico.

### 2.3. Codificación de Variables Categóricas

Vimos que las variables categóricas deben ser etiquetadas como variables numéricas, no como cadenas. En el caso del dataset en cuestión, para tres campos existe una codificación que viene desde la fuente. Por lo tanto, si guardamos estos diccionarios, podemos descartar las columnas de descripción:

In [ ]:
data_uni_sin_ind = data_unique.reset_index()

In [369]:
PermitTypeMap = data_uni_sin_ind[['PermitType','PermitTypeDefinition']].drop_duplicates().sort_values(by = 'PermitType')
PermitTypeMap.to_csv
PermitTypeMap

PermitType                 PermitTypeDefinition
251            1                     new construction
529            2          new construction wood frame
9              3     additions alterations or repairs
6              4                         sign - erect
1562           5  grade or quarry or fill or excavate
43             6                          demolitions
1119           7                 wall or painted sign
0              8               otc alterations permit

In [292]:
data_uni_sin_ind[['ExistingConstructionType','ExistingConstructionTypeDescription']].drop_duplicates().sort_values(
    by = 'ExistingConstructionType').dropna()

ExistingConstructionType ExistingConstructionTypeDescription
3                         1.0                       constr type 1
40                        2.0                       constr type 2
75                        3.0                       constr type 3
207                       4.0                       constr type 4
0                         5.0                      wood frame (5)

In [365]:
data_uni_sin_ind[['ProposedConstructionType','ProposedConstructionTypeDescription']].drop_duplicates().sort_values(
    by = 'ProposedConstructionType').dropna()

ProposedConstructionType ProposedConstructionTypeDescription
3                          1.0                       constr type 1
40                         2.0                       constr type 2
75                         3.0                       constr type 3
1265                       4.0                       constr type 4
0                          5.0                      wood frame (5)

Para el resto de las variables categóricas, utilizaremos un label encoder:

In [366]:
from sklearn import preprocessing

In [367]:
data_unique.describe(include=['O'])

PermitNumber    PermitTypeDefinition   Block     Lot  \
count         198138                  198138  198138  198138   
unique        181298                       8    4896    1055   
top     201602179765  otc alterations permit    3708     001   
freq              68                  178299    1194   10065   

       StreetNumberSuffix StreetName StreetSuffix UnitSuffix   Description  \
count                1941     198138       195371       1924        197855   
unique                 17       1704           21        158        134266   
top                     A     Market           St          A  street space   
freq                 1342       5430       137859        338         22071   

       CurrentStatus CompletedDate StructuralNotification  \
count         198138         96901                   6893   
unique            14          1300                      1   
top         complete    10/30/2015                      Y   
freq           96787           206                   6893   

       VoluntarySoftStoryRetrofit FireOnlyPermit        ExistingUse  \
count                          35          18752             157233   
unique                          1              1                 93   
top                             Y              Y  1 family dwelling   
freq                           35          18752              46737   

              ProposedUse TIDFCompliance ExistingConstructionTypeDescription  \
count              155948              2                              154987   
unique                 94              2                                   5   
top     1 family dwelling              Y                      wood frame (5)   
freq                46324              1                              112876   

       ProposedConstructionTypeDescription SitePermit  \
count                               155225       5316   
unique                                   5          1   
top                         wood frame (5)          Y   
freq                                113895       5316   

       NeighborhoodsAnalysisBoundaries  \
count                           196421   
unique                              41   
top     Financial District/South Beach   
freq                             21780   

                                       Location  
count                                    196446  
unique                                    57604  
top     (37.79226164705184, -122.4034859571375)  
freq                                        554

### 2.4. No Cambiar los Nombres de las Variables de la Fuente de Origen

No modificamos los nombres de las variables de la fuente de origen, ya que no fue necesario.

### 2.5. Verificación de Consistencia de Datos

Este es el paso más analítico, en donde debemos aplicar reglas de integridad.

### 2.6. Identificar y Documentar Valores Atípicos/Outliers

Calcular estadísticos

### 2.7. Evaluar Cómo Comprimir los Datos Para su Almacenamiento Más Eficiente

### 2.8. Guardar el Set de Datos con un Nombre Informativo

## 3. Pasos de Limpieza Deseables

### 3.1. Ordenar Variables/Columnas

Si es posible – primero ID, luego en el mismo orden que la fuente

### 3.2. Quitar Variables/Columnas Irrelevantes

que no Tienen información a analizar

### 3.3. Renombrar Variables de Grillas

### 3.4. Categorizar Resultados en “Otros”

Si tiene un campo de texto libre asociado, codificar en nuevos valores de la variable
categórica asociada. Revisar fuzzyness.

### 3.5. Agregar Metadata a los Datos

Cuando y como fueron obtenidos, limpieza realizada, asunciones, etc. Vincular con etiquetas del código fuente y los datos. Al menos incluir un README.